File: Tree/tree.py
def menu():
    print("+data: make root node")
    print("+data(data): insert child node")
    print("-data: delete node")
    print("=data(datadata): insert sibling")
    print("P(data): get parent")
    print("C(data): get child")
    print("S(data): get sibling")
    print("A(data): get ancestors")
    print("D: get descendants")
    print("L: level of tree / L(data): level of data")
    print("G: degree of tree / G(data): degree of data")
    print("#: count node")
    print("K: clear tree")
    print("T: print tree")
    print("Q: QUIT")
    print("J(newrootrootroot): join tree")
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None  # child
        self.right = None  # sibling


class Tree:
    def __init__(self):
        self.root = None
        self.count = 0
        self.level = 0
        self.degree = 0

    def insert_child(self, parent, item):
        node = Node(item)

        if parent.left is None:
            parent.left = node

        else:
            cur = parent.left
            while cur.right is not None:
                cur = cur.right
            cur.right = node

        self.count += 1

    def insert_sibling(self, sib, item):
        node = Node(item)
        cur = sib
        while cur.right is not None:
            cur = cur.right
        cur.right = node

        self.count += 1

    def print_tree(self, node1, node2):
        cur = node2.left
        if node1 is node2:
            print(node1.data, end='')

        if node1.left is None:
            print()
            return
        print("(", end='')

        while cur is not None:
            print(cur.data, end='')
            if cur.left is not None:
                self.print_tree(node1, cur)
            if cur.right is not None:
                print(",", end='')
            cur = cur.right

        print(")", end='')
        if node1 is node2:
            print()

    def search_node(self, node, item):
        if node is None:
            return
        if node.data == item:
            return node

        cur = self.search_node(node.left, item)
        if cur is not None:
            return cur
        cur = self.search_node(node.right, item)
        if cur is not None:
            return cur

        return None

    def get_parent_bt(self, root, node):
        if root is None:
            return None
        if root.left is node or root.right is node:
            return root
        cur = self.get_parent_bt(root.left, node)
        if cur is not None:
            return cur
        cur = self.get_parent_bt(root.right, node)
        if cur is not None:
            return cur
        return None

    def is_leftchild(self, parent, node):
        if parent.left is node:
            return True
        else:
            return False

    def is_rightchild(self, parent, node):
        if parent.right is node:
            return True
        else:
            return False

    def get_parent(self, root, node):
        if self.get_parent_bt(root, node) is None:
            return None

        cur = node
        while not self.is_leftchild(self.get_parent_bt(root, cur), cur):
            cur = self.get_parent_bt(root, cur)
        return self.get_parent_bt(root, cur)

    def get_child(self, parent):
        if parent.left is None:
            print("NO CHILD")
            return

        cur = parent.left
        while cur is not None:
            print(cur.data, end='')
            cur = cur.right
        print()

    def get_sibling(self, root, node):
        cur = self.get_parent(root, node)
        if cur is None:
            print("{} IS ROOT NODE".format(node.data))
            return
        cur = cur.left
        while cur is not None:
            if cur.data != node.data:
                print(cur.data, end='')
            cur = cur.right
        print()

    def level_of_node(self, root, node):
        level = 0
        cur = node
        while cur is not root:
            cur = self.get_parent(root, cur)
            level += 1
        return level

    def level_of_tree(self, root, node):
        if node is not None:
            self.level_of_tree(root, node.left)
            self.level_of_tree(root, node.right)
            if self.level_of_node(root, node) > self.level:
                self.level = self.level_of_node(root, node)
        return

    def delete_node(self, root, node):
        if node.left is not None:
            print("ERROR")
            return

        parent = self.get_parent_bt(root, node)

        if node.right is None:
            if self.is_leftchild(parent, node):
                parent.left = None
            else:
                parent.right = None

        else:
            child = node.right
            if self.is_leftchild(parent, node):
                parent.left = child
            else:
                parent.right = child

    def get_ancestors(self, node):
        if self.root is node:
            print("NO ANCESTORS")
            return
        cur = node
        while cur is not self.root:
            cur = self.get_parent(self.root, cur)
            print(cur.data, end='')
        print()

    def get_descendants(self, node):
        temp = node.data
        node.data = ''
        self.print_tree(node, node)
        node.data = temp

    def degree_of_node(self, node):
        degree = 0
        cur = node.left
        while cur is not None:
            degree += 1
            cur = cur.right
        return degree

    def degree_of_tree(self, root):
        if root is not None:
            self.degree_of_tree(root.left)
            self.degree_of_tree(root.right)
            if self.degree_of_node(root) > self.degree:
                self.degree = self.degree_of_node(root)
        return

    def count_child(self, node):
        if node is None:
            return -1
        elif node.left == 0:
            return 0
        else:
            num = 0
            cur = node.left
            while cur is not None:
                cur = cur.right
                num += 1
            return num



def copy_tree(root):
    if root is not None:
        newnode = Node(root.data)
        newnode.left = copy_tree(root.left)
        newnode.right = copy_tree(root.right)
        return newnode
    return None


def join_trees(newroot, root1, root2):
    node1 = copy_tree(root1)
    node2 = copy_tree(root2)
    newroot.left = node1
    node1.right = node2
    return newroot


trees = []
idx = -1
menu()
is_binary = int(input("choose binary tree 1/ genaral tree 0: "))
while True:
    enter = input(">> ")
    enter += ' '
    if enter[0] == 'Q':
        break
    i = 0
    while i < len(enter):
        if enter[i] == '+':
            if enter[i + 2] == '(':
                node = trees[idx].search_node(trees[idx].root, enter[i + 1])
                if node is None:
                    print("Not exist")
                else:
                    i += 3
                    while enter[i] != ')':
                        if is_binary == 1 and trees[idx].count_child(node) == 2:
                            print("ERROR")
                            break
                        trees[idx].insert_child(node, enter[i])
                        i += 1
                while enter[i] != ')':
                    i += 1
            else:
                root = Node(enter[i + 1])
                tree = Tree()
                tree.root = root
                trees.append(tree)
                idx = len(trees) - 1
                i += 1
              
        elif enter[i] == '-':
            node = trees[idx].search_node(trees[idx].root, enter[i + 1])
            if node is None:
                print("Not exist")
            else:
                trees[idx].delete_node(trees[idx].root, node)
            i += 1
          
        elif enter[i] == '=':
            node = trees[idx].search_node(trees[idx].root, enter[i + 1])
            if node is None:
                print("Not exist")
            else:
                i += 3
                while enter[i] != ')':
                    if is_binary == 1 and trees[idx].count_child(trees[idx].get_parent(node)) == 2:
                        print("ERROR")
                        break
                    trees[idx].insert_sibling(node, enter[i])
                    i += 1
            while enter[i] != ')':
                i += 1

        elif enter[i] == 'P':
            parent = trees[idx].get_parent(trees[idx].root, trees[idx].search_node(trees[idx].root, enter[i + 2]))
            if parent is None:
                print("ERROR")
            else:
                print(parent.data)
            i += 3
          
        elif enter[i] == 'C':
            if trees[idx].search_node(trees[idx].root, enter[i + 2]) is None:
                print("Not exist")
            else:
                trees[idx].get_child(trees[idx].search_node(trees[idx].root, enter[i + 2]))
            i += 3

        elif enter[i] == 'S':
            node = trees[idx].search_node(trees[idx].root, enter[i+2])
            if node is None:
                print("Not exist")
            else:
                trees[idx].get_sibling(trees[idx].root, node)
            i += 3
          
        elif enter[i] == 'A':
            node = trees[idx].search_node(trees[idx].root, enter[i+2])
            if node is None:
                print("Not exist")
            else:
                trees[idx].get_ancestors(node)
            i += 3
          
        elif enter[i] == 'D':
            node = trees[idx].search_node(trees[idx].root, enter[i+2])
            if node is None:
                print("Not exist")
            else:
                trees[idx].get_descendants(node)
            i += 3
          
        elif enter[i] == 'L':
            if enter[i+1] == '(':
                node = trees[idx].search_node(trees[idx].root, enter[i+2])
                if node is None:
                    print("Not exist")
                else:
                    print("LEVEL: %d" %trees[idx].level_of_node(trees[idx].root, node))
                i += 3
            else:
                trees[idx].level_of_tree(trees[idx].root, trees[idx].root)
                print("LEVEL: %d"%trees[idx].level)
              
        elif enter[i] == 'G':
            if enter[i+1] == '(':
                node = trees[idx].search_node(trees[idx].root, enter[i+2])
                if node is None:
                    print("Not exist")
                else:
                    print("DEGREE: %d"%trees[idx].degree_of_node(node))
                i += 3
            else:
                trees[idx].degree_of_tree(trees[idx].root)
                print("DEGREE: %d"%trees[idx].degree)
              
        elif enter[i] == '#':
            if enter[i+1] == '(':
                node = trees[idx].search_node(trees[idx].root, enter[i+2])
                if node is None:
                    print("Not exist")
                else:
                    print("CHILD NODE: %d"%trees[idx].count_child(node))
                i += 3
            else:
                print("NODE COUNT: %d"%trees[idx].count)
              
        elif enter[i] == 'K':
            if idx == -1:
                print("NO TREE")
            else:
                trees.pop(idx)
                idx -= 1

        elif enter[i] == 'T':
            if idx == -1:
                print("Not exist")
            else:
                trees[idx].print_tree(trees[idx].root, trees[idx].root)

        elif enter[i] == 'J':
            idx1 = -1
            idx2 = -1
            cnt1 = 0
            cnt2 = 0
            for j in range(len(trees)):
                if trees[j].root.data == enter[i+3]:
                    idx1 = j
                    cnt1 = trees[j].count
                elif trees[j].root.data == enter[i+4]:
                    idx2 = j
                    cnt2 = trees[j].count
            if idx1 == -1 or idx2 == -1:
                print("ERROR")
            else:
                newtree = Tree()
                newroot = Node(enter[i+2])
                newtree.root = newroot
                join_trees(newroot, trees[idx1].root, trees[idx2].root)
                newtree.count = cnt1 + cnt2 + 1
                trees.append(newtree)
                idx = len(trees) - 1
            i += 5
        i += 1

File: Tree/tree.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node{
    char data;
    struct Node *left;
    struct Node *right;
} NODE;

void init_node(NODE* node){
    if (node==NULL){
        return;
    }
    node->left=NULL;
    node->right=NULL;
}

void insert_node(NODE *parent, char input){
    NODE * node =(NODE*)malloc(sizeof(NODE));
    node->right=NULL;
    node->left=NULL;
    node->data=input;

    if (parent->left==NULL){
        parent->left=node;
        return;
    }
    else{
        NODE * cur =parent->left;
        while (cur->right!=NULL){
            cur=cur->right;
        }
        cur->right=node;
    }
}

void insert_sibling(NODE *si, char input){
    NODE * node=(NODE*)malloc(sizeof(NODE));
    node->right=NULL;
    node->left=NULL;
    node->data=input;
    NODE *cur=si;
    while (cur->right!=NULL){
        cur=cur->right;
    }
    cur->right=node;
}

void print(NODE *n1, NODE *n2){
    NODE *cur=n2->left;
    if (n1==n2){
        printf("%c",n1->data);
    }
    if (n1->left==NULL){
        printf("\n");
        return;
    }
    printf("(");
    while (cur!=NULL){
        printf("%c",cur->data);
        if (cur->left!=NULL)
            print(n1,cur);
        if (cur->right!=NULL)
            printf(",");
        cur=cur->right;   
    }
    printf(")");
    if (n1==n2){
        printf("\n");
    }
}

NODE* search_node(NODE* root, char input){
    if (root==NULL)
        return NULL;
    if (root->data==input)
        return root;

    NODE *cur;
    cur=search_node(root->left,input);
    if (cur!=NULL)
        return cur;
    cur = search_node(root->right,input);
    if (cur!=NULL)
        return cur;
    return NULL;
}   

int is_leftchild(NODE* parent, NODE* node){
    if (parent->left==node)
        return 1;
    else
        return 0;
}

int is_rightchild(NODE* parent, NODE* node){
    if (parent->right==node)
        return 1;
    else 
        return 0;
}

NODE* get_bparent(NODE *root, NODE *node){
    if (root == NULL)
        return NULL;
    if (root->left==node || root->right==node)
        return root;
    NODE *cur;
    cur=get_bparent(root->left,node);
    if (cur!=NULL)
        return cur;
    cur = get_bparent(root->right,node);
    if (cur!=NULL)
        return cur;
    return NULL;
}

NODE* get_parent(NODE* root, NODE* node){
    if (get_bparent(root,node)==NULL)
        return NULL;
    NODE *cur = node;
    while (is_leftchild(get_bparent(root, cur),cur)!=1){
        cur = get_bparent(root,cur);
    }
    return get_bparent(root,cur);
}

void get_child(NODE* parent){

    if (parent->left==NULL){
        printf("NO CHILD\n");
        return;
    }
    NODE* cur=parent->left;
    printf("{");
    int st=0;
    while (cur != NULL){
        
            if(st!=0) 	printf(",");
			else 		st+=1;
            printf("%c", cur->data);
		
        cur = cur->right;
    }
    printf("}\n");
}

void get_sibling(NODE* root, NODE* node){
    NODE* cur = get_parent(root, node);
    if (cur == NULL){
        printf("%c IS ROOT NODE\n", node->data);
        return;
    }
    cur = cur->left;
	printf("{");
    int st=0;
    while (cur != NULL){
        if (cur->data != node->data)
		{
            if(st!=0) 	printf(",");
			else 		st+=1;
            printf("%c", cur->data);
		}
        cur = cur->right;
    }
    printf("}\n");
}

int level_of_node(NODE* root, NODE* node){
    int l=0;
    NODE* cur=node;
    while (cur!=root){
        cur=get_parent(root,cur);
        l++;
    }
    return l;
}

void level_of_tree(NODE* root, NODE* node, int* l){
    if (node != NULL){
        level_of_tree(root, node->left, l);
        level_of_tree(root, node->right, l);
        if (level_of_node(root, node) > *l)
            *l = level_of_node(root, node);
    }
    return;
}

void delete_node(NODE* root, NODE *node){
    if (node->left!=NULL){
        printf("%c IS PARENT NODE. CANT DELETE.\n",node->data);
        return;
    }

    NODE* parent=get_bparent(root, node);
    if (node->right==NULL){ 
        if(is_leftchild(parent,node)){
            parent->left=NULL;
        }
        else{
            parent->right=NULL;
        }
        free(node);
    }
    else{ 
        NODE *child=node->right;
        if (is_leftchild(parent,node)){
            parent->left=child;
        }
        else{
            parent->right=child;
        }
        free(node);
    }
}

void get_ancestors(NODE* root, NODE* node){
    if (root == node){
        printf("NO ANCESTORS\n");
        return;
    }
    NODE* cur=node;
    while (cur!=root){
        cur=get_parent(root,cur);
        printf("%c ",cur->data);
    }
    printf("\n");
}

void get_descendants(NODE* node){
    char temp = node->data;
    node->data = 8; 
    print(node,node);
    node->data =temp;
}

int degree_of_node(NODE* node){
    int d=0;
    NODE* cur=node->left;
    while(cur!=NULL){
        d++;
        cur = cur->right;
    }
    return d;
}

void degree_of_tree(NODE* root, int* d){
    if (root != NULL){
        degree_of_tree(root->left, d);
        degree_of_tree(root->right, d);
        if (degree_of_node(root) > *d){
            *d =degree_of_node(root);
        }
    }
    return;
}

int count_child(NODE* node){
    if (node==NULL)
        return -1;
    else if (node->left==NULL)
        return 0;
    else{
        int num=0;
        NODE* cur=node->left;
        while (cur!=NULL){
            cur=cur->right;
            num++;
        }
        return num;
    }
}

int count_node(NODE* root){
    if (root!=NULL){
        return 1 + count_node(root->left) + count_node(root->right);
    }
    return 0;
}

NODE* copy_tree(NODE* root){
    if (root != NULL){
        NODE* node = (NODE*)malloc(sizeof(NODE));
        node->data = root->data;
        node->left = copy_tree(root->left);
        node->right = copy_tree(root->right);
        return node;
    }
    return NULL;
}

NODE* join_trees(NODE* newroot, NODE* root1, NODE* root2){
    NODE* node1 = copy_tree(root1);
    NODE* node2 = copy_tree(root2);
    newroot->left = node1;
    node1->right = node2;
    return newroot;
}

void clear(NODE** root, NODE* node){ 
    if (node!=NULL){
        clear(root,node->left);
        clear(root,node->right);
        free(node);
    }
    return;
}

void gt_to_bt(NODE* root) {
  if (root == NULL)
		return;
	if (root->right == NULL && root->left == NULL)
		return;
	printf("(");
	if (root->left != NULL)
		printf("%c", root->left->data);
	gt_to_bt(root->left);
	printf(",");
	if (root->right != NULL)
		printf("%c", root->right->data);
	gt_to_bt(root->right);
	printf(")");
}

void menu(){
  	printf("+data: make root node\n");
    printf("+data(data): insert child node\n");
    printf("-data: delete node\n");
    printf("=+data(data): insert sibling\n");
    printf("P(data): get parent\n");
    printf("C(data): get child\n");
    printf("S(data): get sibling\n");
    printf("A(data): get ancestors\n");
    printf("D(data): get descendants\n");
    printf("L: level of tree\n");
    printf("L(data): level of data\n");
    printf("G: degree of tree\n");
    printf("G(data): degree of data\n");
    printf("#: count node\n");
    printf("J(newrootrootroot): join tree\n");
    printf("K: clear tree\n");
    printf("T: print tree\n");
    printf("Q: QUIT\n");
    printf("B: Change to binary tree\n");
    
}

int main(){
    NODE* roots[10];
    int idx=-1;
    int len=0;
    int ib=0;
    menu();
    printf("choose binary tree 1/ genaral tree 0: ");
    scanf("%d",&ib);
    getchar();
    while (1){
        char input[30];
        printf(">>");
        gets(input);
        if (input[0]=='Q'){
            break;
        }
        int i;
        for (i = 0;input[i]!='\0';i++){
          if(input[i]=='+'){
            if (input[i+2]=='('){
                    NODE* node=search_node(roots[idx], input[i+1]);
                    if (node==NULL){
                        printf("%c is not exist\n",input[i+1]);
                    }
                    else{
                        i += 3;
                        while (input[i]!=')'){
                            if (ib==1&&count_child(node)==2){
                                printf("error\n");
                                break;
                            }
                            insert_node(node,input[i]);
                            i++;
                        }
                    }
                    while (input[i]!=')')
                        i++;
                }
            else{
                    NODE* root = (NODE*)malloc(sizeof(NODE));
                    init_node(root);
                    root->data=input[i+1];
                    idx=len;
                    roots[idx]=root;
                    len++;
                    i += 2;
                }
          }
            else if (input[i]=='T'){
                if (idx==-1){
                    printf("NO TREE\n");
                }
                else{
                    print(roots[idx],roots[idx]);
                }
            }
            else if (input[i]=='-'){
                NODE* node=search_node(roots[idx],input[i+1]);
                if (node==NULL){
                    printf("%c is not exist\n",input[i+1]);
                }
                else{
                    delete_node(roots[idx],node);
                }
                i++;
            }
            else if (input[i]=='='&&input[i+1]=='+'){
                NODE* node=search_node(roots[idx],input[i+2]);
                if (node==NULL){
                    printf("%c is not exist\n",input[i+2]);
                }
                else{
                    i += 4;
                    while (input[i] !=')'){
                        if (ib== 1&&count_child(get_parent(roots[idx],node))==2){
                            printf("error\n");
                            break;
                        }
                        insert_sibling(node,input[i]);
                        i++;
                    }
                }
                while (input[i]!=')')
                    i++;
            }
            else if (input[i]=='P'){
                NODE* parent = get_parent(roots[idx],search_node(roots[idx],input[i+2]));
                if (parent==NULL){
                    printf("error\n");
                }
                else{
                    printf("%c\n",parent->data);
                }
                i += 3;
            }
            else if (input[i]=='C'){
                if (search_node(roots[idx],input[i+2])==NULL){
                    printf("%c is not exist\n",input[i+2]);
                }
                else
                    get_child(search_node(roots[idx],input[i+2]));
                i += 3;
            }
            else if (input[i] == 'S'){
                NODE* node=search_node(roots[idx],input[i+2]);

                if (node == NULL){
                    printf("%c is not exist\n",input[i+2]);
                }
                else{
                    get_sibling(roots[idx],node);
                }
                i += 3;
            }
            else if (input[i]=='A'){
                NODE* node=search_node(roots[idx],input[i+2]);
                
                if (node==NULL){
                    printf("%c is not exist\n",input[i+2]);
                }
                else{
                    get_ancestors(roots[idx],node);
                }
                i += 3;
            }
            else if (input[i]=='D'){
                NODE* node=search_node(roots[idx],input[i+2]);

                if (node==NULL){
                    printf("%c is not exist\n",input[i+2]);
                }
                else{
                    get_descendants(node);
                }
                i += 3;
            }

          else if (input[i]=='L'){
                if (input[i+1]=='('){
                    NODE *node=search_node(roots[idx],input[i+2]);
                    if (node==NULL){
                        printf("%c is not exist\n",input[i+2]);
                    }
                    else{
                        printf("level of node: %d\n",level_of_node(roots[idx],node));
                    }
                    i += 3;
                }
                else{
                    int level=0;
                    level_of_tree(roots[idx],roots[idx],&level);
                    printf("level of tree: %d\n",level);
                }
            }
           else if (input[i]=='G'){
                if (input[i+1]=='('){
                    NODE* node=search_node(roots[idx],input[i+2]);
                    if (node==NULL){
                        printf("%c is not exist\n",input[i+2]);
                    }
                    else{
                        printf("degree: %d\n",degree_of_node(node));
                    }
                    i += 3;
                }
                else{
                    int degree=0;
                    degree_of_tree(roots[idx],&degree);
                    printf("degree: %d\n",degree);
                }
            } 
            else if (input[i]=='#'){
                printf("%d\n",count_node(roots[idx]));
                
            }
            else if (input[i] == 'J'){
                int idx1 = -1;
                int idx2 = -1;
                int j;
                for(j = 0; j < len; j++){
                    if (roots[j]->data == input[i+3])
                        idx1 = j;
                    else if (roots[j]->data == input[i+4])
                        idx2 = j;
                }
                if (idx1 == -1 || idx2 == -1){
                    printf("ERROR\n");
                }
                else{
                    idx = len;
                    NODE* newroot = (NODE*)malloc(sizeof(NODE));
                    init_node(newroot);
                    newroot->data = input[i+2];
                    roots[idx] = join_trees(newroot, roots[idx1], roots[idx2]);
                    len++;
                }
                i += 5;
            }
            else if (input[i]=='K'){
                if (idx==-1){
                    printf("tree is not exist\n");
                }
                else{
                    clear(&roots[idx],roots[idx]);
                    int j;
                    for (j=idx;j < len;j++){
                        roots[j]=roots[j + 1];
                    }
                    len--;
                    if (len==idx){
                        idx--;
                    }
                }    
            }
		//일반트리의 sibling끼리 right로 연결되어 있어 이미 binnary tree형태이기 때문에 출력만 다르게 함 
          else if(input[i]=='B'){//binary tree로 전환
            if(ib==1){
              printf("Already binary tree\n");
            }
            else{
              printf("(%c",roots[idx]->data);
              gt_to_bt(roots[idx]);
              printf(")\n");
            }
          }
        }//for
    }//while
}//main